### ISSUE TITLE
Setting Material in another thread for performance

### ISSUE THREAD

**Issue by AliFlux** [helix-toolkit/helix-toolkit#1022]
## TL;DR

I was profiling my helix/sharpdx application, and apparently setting material to a model takes a lot of CPU. And that can't even be offloaded to a separate thread. Any workaround possible? (see line# 838)

![profiler](https://user-images.githubusercontent.com/8333794/49587009-95431180-f984-11e8-8c5d-42900bc5abcf.PNG)

## Details

I'm working on a project where a lot of new images are being received from network after every time interval. For each image, I've to create a new mesh with that image material attached to it.

The application UI becomes jittery. Upon profiling using Visual Studio Performance Analyzer, it's evident that the UI hangs whenever a material is set to a model. Apparently, it takes a lot of CPU.

Any idea if we could offload this from the UI thread? The [https://github.com/helix-toolkit/helix-toolkit/wiki/Material-Optimization](Material Optimization) page says that it has to be created in UI, but doesn't say anything about setting it to a model.

**Comment by holance** [comment]
This is unfortunate design requirement to be compatible with WPF framework and MVVM, it requires logical tree/visual tree/dependency property update has to be done in UI thread. If you don't need MVVM or XAML, use scene nodes and material core from WPF.SharpDX.Model.Scene namespace. Use SceneNodeGroupModel3D to hold them. This may improve performance a bit without going through the WPF layer.

How many image are you going to receive? If you know a rough limit, try to add a predefined number of models/materials at the beginning and set their `IsRendering = false`. When you receive the image, only update the material diffuse map property and set `IsRendering = true`. For models no longer needed, set `IsRendering = false` instead of removing it. Basically create a predefined model cache to avoid creating new material and model each time a new image is received.

In future, we may consider support async model creation, which may need some architecture change.

**Comment by AliFlux** [comment]
Thanks. `IsRendering` really reduced the effect a bit. Let me play with the architecture and get back on this.

**Comment by AliFlux** [comment]
Is there any example available for `SceneNodeGroupModel3D`, It seems like a new feature.

**Comment by holance** [comment]
This control is new in develop branch, has not been released yet. But you can copy it in your own project. Basically all the Model3D controls in WPF.SharpDX is just a wrapper for the scene nodes. 

You can use scene node very similar to the WPF models. Try to reference to the SharpDX.Core demo, it uses scene node directly.

**Comment by KrabatTilt** [comment]
That is more or less the same problem I tried to bypass by using BatchedMeshGeometryModel3D as mentioned in #1021 . Would be nice to have the SceneNodeGroupModel3D in one of the releases.

**Comment by holance** [comment]
> That is more or less the same problem I tried to bypass by using BatchedMeshGeometryModel3D as mentioned in #1021 . Would be nice to have the SceneNodeGroupModel3D in one of the releases.

You can just grab the [SceneNodeGroupModel3D](https://github.com/helix-toolkit/helix-toolkit/blob/develop/Source/HelixToolkit.SharpDX.SharedModel/Element3D/SceneNodeGroupModel3D.cs) source code and put in your project before next release. The source is actually really simple, just a wrapper of the GroupNode.

**Comment by holance** [comment]
Fixed.
